flomio/passninja-guestlist-gas
# form onboarding, create pass and separation of files to reduce monolithic files

# Changelog
Did WAY more than onboarding, bad habits of not making sure my branches solve one problem each, will keep better attention in the future.

- `createPass_` refactored to new dynamic format
- `onboardNewPassholderFromForm` refactored to new dynamic format and creates a pass as well now.
- Minor changes to `updatePass_` for now until I get more information about the API route.
- documentation revised (added `docs/` folder and all functions are autodocumented).
- created `build.gs` which houses the build sheet/forms functions
- created `config.gs` which houses overall config variables
- implemented a hash lookup/storage to determine whether an update is needed when calling `updateFromConfig_`
- `doGet` no longer returns OAuth as it's unnecessary unless for testing.
- simplified/refactored the doPost route and `addEvent`

## Notes

- `createPass_` is still more monolithic than I'd like.
- dynamic form creation is still limited, not all options are implemented fully, will reflect this by limiting the dropdown in the `Config` sheet.
- Will use this breather as a chance to touch base with @blairexico to check about how we're looking for Friday and testing functionality.

# TO TEST
## REQUIRED SETUP
- Open whichever sheet you have for the Google Sheet + Script
- It MUST have:
1.  A "Config" sheet
1.  After pulling the most recent Script in the Script Editor you **MUST** double check that the `appsscript.json` file (open with `Menu -> View -> Show manifest file`) has these contents:
``` json
{
  "webapp": {
    "access": "ANYONE_ANONYMOUS",
    "executeAs": "USER_DEPLOYING"
  },
  "oauthScopes": ["https://www.googleapis.com/auth/script.container.ui", "https://www.googleapis.com/auth/drive", "https://www.googleapis.com/auth/script.external_request", "https://www.googleapis.com/auth/spreadsheets", "https://www.googleapis.com/auth/forms", "https://www.googleapis.com/auth/script.scriptapp"],
  "runtimeVersion": "V8"
}
```
3.  That sheet must have two Named ranges that have similar fields to this:
![image](https://user-images.githubusercontent.com/1587270/74388368-3fa89700-4dc9-11ea-9992-ea0230a50d1a.png)

- `config_passTypeId` - The blue `demo.testing` field is the named range (single cell)
- `config_fields` - The values underneath the Field header.  In Pass? is either blank or "Y", if the field is "In Pass: Y" then it must match the field that the passtype requires.  The form-type is unnecessary to add for now, it defaults to a text field if you do not put anything there, but the config_fields must be 4 columns wide.
![image](https://user-images.githubusercontent.com/1587270/74388533-9f9f3d80-4dc9-11ea-8dfa-8a90dfe9ad10.png)

## To Verify Sheet/Form creation
1.  You can test it by opening the spreadsheet and going to the PassNinja menu and running the "Build/Update From Config" menu item.
1.  This will build the Sheets: `Form Responses #`, `Events`, `Contacts` and the connected Form (`Menu -> Form -> Edit Form` can help you verify that the form is correct and matches the fields you entered in the `Config` sheet.)

## To Verify Form Submission
1.  Check with `Menu -> Form -> Go to Live Form`
1.  Enter all the required field information (I made sure all "In Pass?: Y" fields are set as required)
1.  When it is submitted it will appear on the `Form Responses` sheet and also a new contact will be added to the `Contacts` sheet.  
1.  This contact will ALSO have an automatically generated pass, so if you sit tight a couple seconds it will appear.

## To Verify POST/GET requests
1.  Once you deploy the Script in the `Script Editor` using `Publish -> Deploy as Web App` and create a new version you should copy the resulting url.
1.  You can enter this Google Sheets API URL as the cognito user callback url (more information once I talk with @blairexico about this) and it should post the event stream for that pass type to the Events sheet.  See similar instructions here: https://github.com/flomio/passninja-api/pull/238
1.  Now you can feel free to manually do POST or GET requests at that url.  A GET request, if you add `?serialNumber=<YOUR SERIAL HERE>` will return the matching sheet entry that has that serial number in the Contacts sheet `serialNumber` column. The GET request I've been using looks like:
``` bash
curl --location --request GET 'https://script.google.com/macros/s/AKfycby9as7cTqifeqxU8o5_Hl4Rdy4t7vLLjVw_ZCYvqMsqI3Df8l3I/exec?serialNumber=random'
```
4.  The POST request that I've been sending resembles this:
``` bash
curl --location --request POST 'https://script.google.com/macros/s/AKfycby9as7cTqifeqxU8o5_Hl4Rdy4t7vLLjVw_ZCYvqMsqI3Df8l3I/exec' \
--header 'Content-Type: application/json' \
--data-raw '{
    "callback": "https://script.google.com/macros/s/asdf/dev",
    "date": "2020-02-08T15:15:13.341Z",
    "event": {
      "passJson":
        "{\"formatVersion\":1,\"description\":\"Pass Entry for Employee Number 232\",\"passTypeIdentifier\":\"pass.com.passninja.passentry.demo\",\"organizationName\":\"PassEntry\",\"teamIdentifier\":\"Q338UYGFZ8\",\"serialNumber\":\"810ce8e0-1f11-4e29-a35e-cdf326fb90c2\",\"backgroundColor\":\"rgb(41, 50, 58)\",\"labelColor\":\"rgb(255, 132, 119)\",\"foregroundColor\":\"rgb(255, 255, 255)\",\"logoText\":\"PassEntry\",\"generic\":{\"headerFields\":[{\"key\":\"passId\",\"label\":\"PASS ID\",\"value\":\"232\"}],\"primaryFields\":[{\"key\":\"staffName\",\"label\":\"NAME\",\"value\":\"Nico Cary\"}],\"secondaryFields\":[{\"key\":\"jobRole\",\"label\":\"JOB ROLE\",\"value\":\"Chief Innovation Officer\"},{\"key\":\"passExpiry\",\"label\":\"PASS EXPIRY\",\"value\":\"23 December 2020\"}]},\"nfc\":{\"message\":\"810ce8e0-1f11-4e29-a35e-cdf326fb90c2\",\"encryptionPublicKey\":\"MDkwEwYHKoZIzj0CAQYIKoZIzj0DAQcDIgADTkFu6xr9i4GKk1+Jn+UxayuuC+SQLfXA0sE5J5iQx3o=\"},\"webServiceURL\":\"https://api.passninja.com/thumbnail\",\"authenticationToken\":\"8e98126bb7e60666\"}",
      "passType": "passentry.demo",
      "serialNumber": "810ce8e0-1f11-4e29-a35e-cdf326fb90c2",
      "type": "APPLE_CREATE"
    },
    "id": "#niko@passentry.com#passentry.demo#810ce8e0-1f11-4e29-a35e-cdf326fb90c2"
}'
```

## To Verify Activating Pass Creation from the spreadsheet:
1.  In the Contacts sheet, feel free to click anywhere in a row for a specific contact that has all the required fields filled out.
1.  Go to `Menu -> PassNinja -> Create A Pass` and the script should run.
1.  You will see the resulting pass + pass information output into the  contact's fields.

## To Verify Rebuilding from Config in an existing built spreadsheet:
1.  You can add a field to the Config sheet or remove one (just press delete while highlighting the four columns in the row).
1.  Go to `Menu -> PassNinja -> Build/Update from Config` and it should run.
1.  The `Events` sheet will remain untouched
1.  The `Contacts` sheet headers will change but the data below will remain untouched (so the user can feel free to reorder them to fit the new headers and also not lose existing passes.)
1.  The `Form Responses` sheet will now be renamed to `Form Responses {DATE/TIME}` to indicate it is archived and a new `Form Responses` sheet will be created.
1.  The linked Form itself will be updated to the new fields specified in the `Config` sheet.
